src/Compression/CompressionCodecEncrypted.cpp

```diff
@@ -186,6 +191,138 @@ size_t decrypt(std::string_view ciphertext, char * plaintext, EncryptionMethod m
+/// This function get key and nonce and encrypt text with their help.
+/// If something went wrong (can't init context or can't encrypt data) it throws exception.
+/// It returns length of encrypted text.
+size_t decrypt(std::string_view ciphertext, char * plaintext, EncryptionMethod method, const String & key, const String & nonce)
+{
+    EVP_CIPHER_CTX *decrypt_ctx;
+
+    if (!(decrypt_ctx = EVP_CIPHER_CTX_new()))
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    const int ok_cryptinit = EVP_DecryptInit_ex(decrypt_ctx,
+                                                getMethod(method)(),
+                                                nullptr, nullptr, nullptr);
+    if (!ok_cryptinit)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    const int ok_cipherctrl = EVP_CIPHER_CTX_ctrl(decrypt_ctx,
+                                                  EVP_CTRL_GCM_SET_IVLEN,
+                                                  static_cast<int32_t>(nonce.size()), nullptr);
+    if (!ok_cipherctrl)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    const int ok_nonceinit = EVP_DecryptInit_ex(decrypt_ctx, nullptr, nullptr,
+                                                reinterpret_cast<const uint8_t*>(key.data()),
+                                                reinterpret_cast<const uint8_t *>(nonce.data()));
+    if (!ok_nonceinit)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    size_t out_len;
+    const int ok_decryptudpate = EVP_DecryptUpdate(decrypt_ctx,
+                                                   reinterpret_cast<uint8_t *>(plaintext),
+                                                   reinterpret_cast<int32_t *>(&out_len),
+                                                   reinterpret_cast<const uint8_t *>(ciphertext.data()),
+                                                   static_cast<int32_t>(ciphertext.size()) - tag_size);
+    if (!ok_decryptudpate)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    const int ok_tag = EVP_CIPHER_CTX_ctrl(decrypt_ctx,
+                                           EVP_CTRL_GCM_SET_TAG,
+                                           tag_size,
+                                           reinterpret_cast<uint8_t *>(const_cast<char *>(ciphertext.data())) + ciphertext.size() - tag_size);
+    if (!ok_tag)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    const int ok_decryptfinal = EVP_DecryptFinal_ex(decrypt_ctx,
+                                                    reinterpret_cast<uint8_t *>(plaintext) + out_len,
+                                                    reinterpret_cast<int32_t *>(&out_len));
+    if (!ok_decryptfinal)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    EVP_CIPHER_CTX_free(decrypt_ctx);
+
+    return ciphertext.size() - tag_size;
```

Comment:
the same as above - why we don't use `out_len`? - isn't it more reliable?

Reply:
same as above:
https://github.com/google/boringssl/blob/master/crypto/fipsmodule/cipher/aead.c#L219